Fixes #32752 - Don't allow editing readonly settings

tbrisker opened the pull request:
<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

tbrisker committed:
Fixes #32752 - Don't allow editing readonly settings

amirfefer commented on webpack/assets/javascripts/react_app/components/SettingsTable/components/SettingCellInner.js:
```diff
@@ -14,7 +14,11 @@ const SettingCellInner = props => {
+    <span
+      {...rest}
+      className={cssClasses}
+      onClick={() => !setting.readonly && onEditClick(setting)}
```
This still adds a redundant onClick in read-only fields.
How about using a conditional prop rendering? 

```js
  const onClickHandler = {
    ...(!setting.readonly && { onClick: () => onEditClick(setting) }),
  };

  const field = (
    <span {...rest} className={cssClasses} {...onClickHandler}>
      {valueToString(setting)}
    </span>
  );
```

tbrisker replied:
Thanks for the suggestion! I actually took it one step further in the second commit and moved the entire edit handling into the cell, lmk what you think about it (i'm guessing tests/lint will fail, but didn't want to invest in fixing them before getting feedback on the general approach)

tbrisker committed:
Refs #32752 - Move setting editing handling into the cell

Instead of having the table handle the modal and passing down the
callback to trigger it, move the editing logic into the inner cell.

amirfefer reviewed (commented):
I think this approach simplifies the code a bit, it doesn't make sense to pass the `onEditClick` between nodes.  calling `useDispatch` in a leaf component doesn't affect performance either.

Can you please just fix the failures?

amirfefer commented on webpack/assets/javascripts/react_app/components/SettingsTable/components/SettingCellInner.js:
```diff
@@ -2,19 +2,36 @@ import React from 'react';
   });
 
+  if (!setting.readonly){
+    const { setModalOpen } = useSettingModal();
```
hooks cannot be called under conditions

tbrisker replied:
thanks, fixed

amirfefer commented on webpack/assets/javascripts/react_app/components/SettingsTable/components/SettingCellInner.js:
```diff
@@ -2,19 +2,36 @@ import React from 'react';
+    const { setModalOpen } = useSettingModal();
+    const dispatch = useDispatch();
+    const onEditClick = async setting => {
+      await dispatch(setSettingEditing(setting));
```
It isn't related, but `setSettingEditing` isn't async, I wonder why `onEditClick` needs to be asynchronous.

tbrisker replied:
looks like it doesn't, fixed

tbrisker committed:
Refs #32752 - simplify setSettingEditing action

There is no need for the action to be async

tbrisker committed:
Refs #32752 - fix eslint

tbrisker committed:
Refs #32752 - update snapshots

tbrisker commented:
Updated, I broke my changes down to multiple commits to make the review easier, please squash on merge.

tbrisker commented:
JS test failure unrelated, should be fixed by GH-8627

amirfefer commented on webpack/assets/javascripts/react_app/components/SettingsTable/components/SettingCellInner.js:
```diff
@@ -2,19 +2,32 @@ import React from 'react';
+      dispatch(setSettingEditing(setting));
+      setModalOpen();
+    };
+  }
```
nit: This page will need a PF4 refactoring some day, so I'd say we need to keep it a little more explicit :)

```js
  const canClick = !setting.readonly;
  const openModal = () => {
    dispatch(setSettingEditing(setting));
    setModalOpen();
  };

  const field = (
    <span
      // when `onClick` is undefined  react extracts it from the DOM.
      onClick={canClick ? openModal : undefined}
      {...rest}
      className={cssClasses}
    >
      {valueToString(setting)}
    </span>
  );
```

amirfefer reviewed (commented):
Thanks @tbrisker 👍 Tested and works well, just a small nitpick

tbrisker committed:
Refs #32752 - Explicitly add onClick listener

tbrisker commented:
@amirfefer updated with explicit onclick listener. Don't forget to squash on merge please!

amirfefer reviewed (approved):
Thanks @tbrisker !
